### PR TITLE
improved unitful where

### PR DESCRIPTION
**Pull request by sIHURs**
Imprvoed unitful where function, tackles issue https://github.com/ymahlau/fdtdx/issues/191

### CODE REVIEW

File: src/fdtdx/functional/numpy.py [hunk 1]
```diff
@@ -1416,7 +1416,14 @@ def where(
     assert condition.unit.dim == {} and condition.unit.scale == 0, f"Invalid condition input: {condition}"
     x_align, y_align = align_scales(x, y)
     c_val = condition.val
-    new_val = jnp._orig_where(c_val, x_align.val, y_align.val, *args, **kwargs)  # type: ignore
+
+    if any(isinstance(v, jax.Array) for v in [x_align.val, y_align.val, c_val]):
+        new_val = jnp._orig_where(c_val, x_align.val, y_align.val, *args, **kwargs)  # type: ignore
+    elif all(isinstance(v, np.ndarray) for v in (x_align.val, y_align.val, c_val)):
```
Comment (ymahlau) [inline src/fdtdx/functional/numpy.py]: I believe this should be als 'any', not 'all', since there may be python floats / ints / bool etc as input

File: src/fdtdx/functional/numpy.py [hunk 2]
```diff
@@ -1425,42 +1432,53 @@ def where(
...
             new_static_arr = np.where(c_arr, x_arr, y_arr, *args, **kwargs)
-    return Unitful(val=new_val, unit=x.unit, static_arr=new_static_arr)
+    return Unitful(val=new_val, unit=x_align.unit, static_arr=new_static_arr)
 
 
 @overload
 def where(
-    condition: jax.Array | Unitful,
+    condition: ArrayLike | Unitful,
     x: ArrayLike,
```
Comment (ymahlau) [inline src/fdtdx/functional/numpy.py]: Here x can also be ArrayLike | Unitful. The pattern should be that if at least one input is unitful, the all other other inputs can be whatever and the output is unitful

File: src/fdtdx/functional/numpy.py [hunk 3]
```diff
@@ -1425,42 +1432,53 @@ def where(
...
-    return where(c, Unitful(val=x, unit=EMPTY_UNIT), y, *args, **kwargs)
+    return where(c, Unitful(val=x, unit=EMPTY_UNIT), y_non_dim, *args, **kwargs)
 
 
 @overload
 def where(
-    condition: jax.Array | Unitful,
+    condition: ArrayLike | Unitful,
     x: Unitful,
     y: ArrayLike,
```
Comment (ymahlau) [inline src/fdtdx/functional/numpy.py]: Similarly, y can be ArrayLike | Unitful here

File: src/fdtdx/functional/numpy.py [hunk 4]
```diff
@@ -1425,42 +1432,53 @@ def where(
...
+    x_non_dim = Unitful(val=x.val, unit=new_unit)
     c = condition if isinstance(condition, Unitful) else Unitful(val=condition, unit=EMPTY_UNIT)
-    return where(c, x, Unitful(val=y, unit=EMPTY_UNIT), *args, **kwargs)
+    return where(c, x_non_dim, Unitful(val=y, unit=EMPTY_UNIT), *args, **kwargs)
 
 
 @overload
 def where(
-    condition: jax.Array,
+    condition: ArrayLike,
```
Comment (ymahlau) [inline src/fdtdx/functional/numpy.py]: Here we need more fine-grained functions. Please add four functions for jax.Array output similar to the unitful functions above (one functions with all inputs jax.Array and three functions where one of the inputs is definitely jax.Array and the other inputs can be whatever). Then add one function for numpy / python input that returns a numpy array.

Comment (sIHURs) [reply src/fdtdx/functional/numpy.py]: thanks for the comments. I split the function into more fine-grained overloads. But I didn’t add a separate overload for the case where all inputs are jax.Array, because that caused overlapping overload errors. The current three functions return jax.Array and the one returns np.ndarray  I think cover inputs type.

File: src/fdtdx/functional/numpy.py [hunk 5]
```diff
@@ -1425,31 +1437,62 @@ def where(
...
+    if isinstance(condition, Unitful) and isinstance(x, Unitful):
+        return where(condition, x, y, *args, **kwargs)
+
     c = condition if isinstance(condition, Unitful) else Unitful(val=condition, unit=EMPTY_UNIT)
-    return where(c, Unitful(val=x, unit=EMPTY_UNIT), y, *args, **kwargs)
+    new_x = x if isinstance(x, Unitful) else Unitful(val=x, unit=EMPTY_UNIT)
+    new_y = (
+        y
+        if isinstance(x, Unitful)
+        else Unitful(val=y.val, unit=Unit(scale=y.unit.scale, dim={}), static_arr=y.static_arr)
```
Comment (ymahlau) [inline src/fdtdx/functional/numpy.py]: this if else is a bit superficial and can actually lead to bugs. The correct way would be to just use y always I think

File: src/fdtdx/functional/numpy.py [hunk 6]
```diff
@@ -1425,31 +1437,62 @@ def where(
...
+    if isinstance(y, Unitful) and isinstance(condition, Unitful):
+        return where(condition, x, y, *args, **kwargs)
+
     c = condition if isinstance(condition, Unitful) else Unitful(val=condition, unit=EMPTY_UNIT)
-    return where(c, x, Unitful(val=y, unit=EMPTY_UNIT), *args, **kwargs)
+    new_y = y if isinstance(y, Unitful) else Unitful(val=y, unit=EMPTY_UNIT)
+    new_x = (
+        x
+        if isinstance(y, Unitful)
+        else Unitful(val=x.val, unit=Unit(scale=x.unit.scale, dim={}), static_arr=x.static_arr)
```
Comment (ymahlau) [inline src/fdtdx/functional/numpy.py]: same here

File: tests/functional/test_numpy.py [hunk 7]
```diff
@@ -1113,3 +1115,137 @@ def test_arctanh_unitful_with_jax_array():
...
+        if is_currently_compiling() and not unitful.STATIC_OPTIM_STOP_FLAG:
+            assert out.static_arr is not None
+        return out
+
+    jitted_fn = jax.jit(fn)
+    result = jitted_fn(c, x, y)
+
+    expected = np.where(np.array([True, False, True]), np.array([1.0, 2.0, 3.0]), np.array([10.0, 20.0, 30.0]))
+
+    assert jnp.allclose(result.value(), jnp.array(expected))
```
Comment (ymahlau) [inline tests/functional/test_numpy.py:1247]: Can you add another test case which tetsts that invalid inputs also lead to throwing an error? For example, test that an exception is thrown if x and y have different units

File: src/fdtdx/functional/numpy.py (hunk 5 above)
Comment (sIHURs) [reply src/fdtdx/functional/numpy.py]: There’s a question here: we need to keep the dims of x and y consistent.
When an ArrayLike input (for example x) is promoted to Unitful, should we assign x the unit.dim of the other Unitful input (for example y), or should we instead strip the unit.dim from y?

Comment (ymahlau) [reply src/fdtdx/functional/numpy.py]: This is exactly what I am talking about. If the unit.dim are different then an Exception should be thrown. I believe you can just let the units as is and let the error handling in the other unitful where definition take care of it
